Development

avladimirov opened the pull request:
Base classes and solution structure
Lots of utility operations
Code not yet tested

Commit:
- initial solution setup
- update & clean of packages
- added classes for entity models
- external operations separated to utility class
- Implemented baseline client with abstraction to common processing

Commit:
Updated git ingore with VS internal folder & packages folder

aarondcoleman commented on AzureAPIManagement/Utilities/Constants.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        internal static class ApiManagement
+        {
+            internal static readonly string AccessToken = "SharedAccessSignature";
```
Needs to be parameterized. We may publish this within different accounts / staging environments with automatic transforms

aarondcoleman commented on AzureAPIManagement/Services/ApiManagementClientBase.cs:
```diff
@@ -0,0 +1,129 @@
+            // https://msdn.microsoft.com/library/azure/5b13010a-d202-4af5-aabf-7ebc26800b3d#ProgrammaticallyCreateToken
+            // id - the value from the identifier text box in the credentials section of the
+            //      API Management REST API tab of the Security section.
+            string id = "<your identifier value here>";
```
Needs to be parameterized and set on constructor. Environments will determine which keys are used (we have development and production for example).

aarondcoleman commented on AzureAPIManagement/Services/ApiManagementClientBase.cs:
```diff
@@ -0,0 +1,129 @@
+            string id = "<your identifier value here>";
+
+            // key - either the primary or secondary key from that same tab.
+            string key = "<either the primary or secondary key here>";
```
Needs to be parameterized and set on constructor. Environments will determine which keys are used (we have development and production for example).

aarondcoleman commented on AzureAPIManagement/Services/ApiManagementClientBase.cs:
```diff
@@ -0,0 +1,129 @@
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        protected HttpRequestMessage BuildRequest(object data = null)
+        {
+            StackTrace st = new StackTrace();
```
This concerns me. I'd like to avoid Reflection code. Please implement this another way. If instead of an attribute you just have to hardcode this as a string in the method to pass as a parameter, that's fine. It's inline in the attribute anyways, no worse moving it a couple lines down.

Commit:
folder rename

Commit:
Products REST - added remaining methods
Omitted usage of reflection & attributes for mapping rest urls

Commit:
Extended data model with Group entity and abstraction on common properties

Commit:
Files reorgniazation and renames
Test project replaced the console
Test methods added to validate data model deserialization
API updates to accomodate quiery filters and custom results

Commit:
Argument validatation added to client

Commit:
Client configuration via arguments in constructor
Cancellation support for async client methods
Error data model added

Commit:
Fixes , renames and additional tests to client calls

Commit:
Model extended to handle error cases
Client methods describe the expected success response code
Put/Delete operations override in all cases

Commit:
Base class for result of operations with no data retrieval

Commit:
Client - exception handling revised
Data Model - simplified, minor improvements
Tests - added to produc CRUD operations

Commit:
Fixed filtering and building query params

Commit:
Simplifying collection results

Commit:
Fixes and test to API methods

Commit:
Fixes to Product API methods
Added back initial operations - commented temporarily

Commit:
Fixes and test to Product policy operations

Commit:
Implementation and tests for User operations - create & get SSO url

Commit:
Implementation and tests added for Product/User Subscriptions support

Commit:
Minor cleanup
